### ISSUE TITLE
FR: Setting to avoid full refresh on pages with images

### ISSUE THREAD

**Issue by sargunv** [koreader/koreader#10047]
**Does your feature request involve difficulty completing a task? Please describe.**

I have KOReader set to Full Refresh = Never, disabled "Always flash on chapter boundaries", and enabled "Avoid mandatory black flashes in UI", but still KOReader will flash on any page that happens to include an image. This is quite common on the books I read, as Brandon Sanderson's books tend to have chapter icons at every chapter boundary.

My device is pretty good at page turning without ghosting, so I'd prefer to not have KOReader force a full refresh at all as it de-immerses me from the book at the start of every chapter. I'm happy to force a refresh manually when the image is complex (such as a map or photo in the book).

**Describe the solution you'd like**

I'd like either the "Never" setting for Full Refresh to actually flash never, or a separate setting that disables the full refresh when a page turn includes an image.

**Describe alternatives you've considered**

I guess the alternative is for this setting to not exist? and for this flash to be mandatory as it is now.

A more complex alternative: somehow only flash on large images (like full screen images) but not on small ones (like chapter icons). This seems too complex to be worth it though, especially if it's configurable. I'm fine with a manual refresh on pages with large images.

**Additional context**

Unsure if this is KOReader behavior or system behavior, but I think it's KOReader because the system's default NeoReader app doesn't necessarily flash pages with images.

My device: Boox Leaf 2 (Android, E Ink Carta Plus screen)

**Comment by poire-z** [comment]
> A more complex alternative: somehow only flash on large images (like full screen images) but not on small ones (like chapter icons). This seems too complex to be worth it though

Of course, @NiLuJe went this complex way :) - for EPUBs, crengine computes the number and amount of images:

https://github.com/koreader/koreader/blob/d2f344becf3dddc011e086f7a13a609c0e0db752/frontend/apps/reader/modules/readerview.lua#L246-L256

A 7.5% (it seems) is hardcoded, but you could try tweaking that in the code. (Dunno how large are these "chapter icons".)
Then, we could have this ratio tunable by a hidden setting, or have a setting in the eInk menu to allow always avoiding them.
You can also user-patch `Document:getDrawnImagesStatistics()` to return always `nil, nil`.

**Comment by NiLuJe** [comment]
FWIW, the heuristics is "tuned" to skip over graphical *scene breaks*, mostly. Chapter headings are... way more variable, and, given the default "flash on chapter boundaries" behavior, it wouldn't really make sense to even attempt to deal with.

TL;DR: I'm not opposed to a dedicated toggle àla chapter boundaries for this, but I don't think tweaking the heuristics would be a great approach ;).

**Comment by NiLuJe** [comment]
(Also, keep in mind you're using an Android device, where flashing refreshes might be delayed. On a sane device, flashing refreshes are really really fast ;)).

**Comment by sargunv** [comment]
> FWIW, the heuristics is "tuned" to skip over graphical scene breaks, mostly. Chapter headings are... way more variable, and, given the default "flash on chapter boundaries" behavior, it wouldn't really make sense to even attempt to deal with it.

I see the existing logic enables dithering AND a full refresh if image coverage is > 7.5%. I haven't experimented yet, but I'm _guessing_ I want dithering still on chapter icons, just not the full refresh. 

Would you be opposed to a a configurable "image coverage for full refresh" threshold separate from the 7.5% dithering threshold? Personally I'd probably want it at ~50%, targeting covers and other full page images. Or should I just implement a boolean toggle to disable this refresh behavior?

Is the full refresh somehow required for dithering, such that the thresholds need to remain the same, or the boolean toggle would _also_ need to disable dithering? I'm _guessing_ no, because the full refresh happens "only if it's the first time we're painting it".

> Also, keep in mind you're using an Android device, where flashing refreshes might be delayed. On a sane device, flashing refreshes are really really fast ;)

That's fair, though I haven't noticed a difference between the speed of flashing refreshes on my Kindle vs my Boox. I _have_ noticed though my Kindle does normal page changes way faster than my Boox does (in either NeoReader or KOReader). The Kindle is pretty much instant, while the Boox takes a few hundred ms, slightly more in KOReader vs NeoReader.

I replaced the Kindle because I can't jailbreak the current fw version. I went with Onyx Boox over Kobo b/c I use a VPN to remotely access my library of books on my home network, rather than exposing my NAS directly to the internet, and afaik I can't run WireGuard on Kobo. I might pick up a Kobo (or PocketBook?) to try it out though; if reading on a non-Android device is that much better then I may just go ahead and expose my NAS's OPDS feed over the internet so I don't need the VPN.

**Comment by sargunv** [comment]
Ah, I see [a >100 line comment](https://github.com/koreader/koreader/blob/d2f344becf3dddc011e086f7a13a609c0e0db752/frontend/ui/uimanager.lua#L405-L532) on `UIManager:setDirty`. Haven't read it yet, but I'm guessing it'll answer some or all of my questions 😄

**Comment by pazos** [comment]
> (Also, keep in mind you're using an Android device, where flashing refreshes might be delayed. On a sane device, flashing refreshes are really really fast ;)).

Indeed, up to [1/4 seconds](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/device/epd/qualcomm/QualcommEPDController.kt#L18) on onyx.

We do this because we cannot disable system updates and if we race with the system we always win so the full refresh we trigger gets messed up by the partial refresh the system always do.

**Comment by NiLuJe** [comment]
> Ah, I see [a >100 line comment](https://github.com/koreader/koreader/blob/d2f344becf3dddc011e086f7a13a609c0e0db752/frontend/ui/uimanager.lua#L405-L532) on `UIManager:setDirty`. Haven't read it yet, but I'm guessing it'll answer some or all of my questions smile

You definitely don't want to touch the `dithered` flag, and just make the `setDirty` call conditional, that's the only thing requesting a flash here ;).

**Comment by NiLuJe** [comment]
> Would you be opposed to a a configurable "image coverage for full refresh" threshold separate from the 7.5% dithering threshold? Personally I'd probably want it at ~50%, targeting covers and other full page images. Or should I just implement a boolean toggle to disable this refresh behavior?

That feels slightly too nerdy and complicate to tune for to make it actually useful, yeah, I'd rather live with a boolean flag ;).

**Comment by sargunv** [comment]
> We do this because we cannot disable system updates and if we race with the system we always win so the full refresh we trigger gets messed up by the partial refresh the system always do.

I see, so more reason for me to just disable KOReader's full refreshes and rely on the system's (I _think_ I can configure the Boox system to do a refresh on every N interactions)

**Comment by sargunv** [comment]
Filed a preliminary PR for the boolean toggle: https://github.com/koreader/koreader/pull/10049

Haven't tested on actual e-ink as I'm unsure how to get the dev build onto my Android device.

**Comment by Frenzie** [comment]
Fixed by  #10049 I believe.
